server/src/main/java/io/druid/client/DruidDataSource.java

```diff
@@ -128,52 +126,20 @@ public ImmutableDruidDataSource toImmutableDruidDataSource()
       return new ImmutableDruidDataSource(
           name,
           ImmutableMap.copyOf(properties),
-          ImmutableMap.copyOf(partitionNames),
+          ImmutableMap.copyOf(idToSegmentMap),
           ImmutableSet.copyOf(segmentsHolder)
       );
     }
   }
 
   @Override
-  public boolean equals(Object o)
+  public String toString()
   {
-    if (this == o) {
-      return true;
-    }
-    if (o == null || getClass() != o.getClass()) {
-      return false;
-    }
-
-    DruidDataSource that = (DruidDataSource) o;
-
-    if (name != null ? !name.equals(that.name) : that.name != null) {
-      return false;
-    }
-    if (partitionNames != null ? !partitionNames.equals(that.partitionNames) : that.partitionNames != null) {
-      return false;
-    }
-    if (properties != null ? !properties.equals(that.properties) : that.properties != null) {
-      return false;
-    }
-    if (segmentsHolder != null ? !segmentsHolder.equals(that.segmentsHolder) : that.segmentsHolder != null) {
-      return false;
+    synchronized (lock) {
+      return "DruidDataSource{" +
+             "properties=" + properties +
+             ", partitions=" + segmentsHolder.toString() +
+             '}';
     }
-
-    return true;
-  }
-
-  @Override
-  public int hashCode()
-  {
-    int result = name != null ? name.hashCode() : 0;
```

Comment:
(optional) an alternative could be to throw `UOE` to catch any bad uses of this class.

Reply:
Thanks. Added.